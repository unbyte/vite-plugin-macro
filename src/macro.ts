import traverse, { Node, NodePath } from '@babel/traverse'
import * as types from '@babel/types'
import { parse, parseExpression } from '@babel/parser'
import template from '@babel/template'
import { CallExpression, ImportDeclaration, Program } from '@babel/types'
import { validateFnName } from './common'

export type Context = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /*
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath
}

export type BabelTools = {
  types: typeof types
  template: typeof template
  traverse: typeof traverse
  parse: typeof parse
  parseExpression: typeof parseExpression
}

/**
 * ```typescript
 * import 'moduleName'
 * { moduleName: string }
 *
 * import defaultName from 'moduleName'
 * { defaultName: string; moduleName: string }
 *
 * import { exportName } from 'moduleName'
 * { exportName: string; moduleName: string }
 *
 * import { exportName as localName } from 'moduleName'
 * { localName: string; exportName: string; moduleName: string }
 *
 * import * as namespaceName from 'moduleName'
 * { namespaceName: string; moduleName: string }
 * ```
 */
export type ImportOption =
  | { moduleName: string }
  | { defaultName: string; moduleName: string }
  | { localName: string; exportName: string; moduleName: string }
  | { exportName: string; moduleName: string }
  | { namespaceName: string; moduleName: string }

export type Helper = {
  /**
   * Get the directory of the root project or the nearest project.
   * @param which 'root' or 'leaf'
   */
  projectDir: (which: 'root' | 'leaf') => string

  /**
   * Normalize path pattern so can resolve file paths as import paths
   * @param path relative path or path pattern
   * @param root absolute file path of the project dir, default to projectDir('leaf')
   * @param importer absolute file path of the importer, default to current module
   *
   * e.g. search files by glob pattern and then import them
   *
   * ```typescript
   * // /src/example/a.ts
   * // /src/example/b.ts
   * // in /src/another/index.ts
   * const pattern = '../example/*.ts'
   * const { normalized, base, resolveImportPath } = normalizePathPattern(pattern)
   * const importPaths = glob.sync(normalized, { cwd: base }).map(resolveImportPath)
   *
   * // importPaths = ['../example/a.ts', '../example/b.ts']
   * ```
   */
  normalizePathPattern: (
    path: string,
    root?: string,
    importer?: string
  ) => {
    /**
     * Normalized path or path pattern,
     */
    normalized: string
    /**
     * Base path for the normalized path or path pattern.
     */
    base: string
    /**
     * Get import path from file path got by normalized path and base,
     * almost the opposite operation of normalization
     * @param path file path
     */
    resolveImportPath: (path: string) => string
  }

  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration>

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration>

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not trim extension for moduleName.
   * @param imports an import
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not trim extension for moduleName.
   * @param imports an import
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (imports: ImportOption, program?: NodePath<Program>) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Get current program or the program of provided node.
   * @param node any node path
   */
  getProgram: (node?: NodePath) => NodePath<Program>

  /**
   * Force to recollect imported macros in the next loop.
   * Usually used after importing new macros to program during expanding a macro.
   *
   * Note:
   *
   *  Only collected imported macros can be applied, because plugin only
   *  searches call expressions for collected macros. Also, external macros
   *  cannot be collected because plugin is unaware of external macros.
   */
  forceRecollectMacros: () => void
}

export type MacroHandler = (
  ctx: Context,
  babel: Readonly<BabelTools>,
  helper: Readonly<Helper>
) => void

type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  typeDefinitions: string[]
}

export type Macro = {
  name: string
  meta: MacroMeta
  apply: MacroHandler
}

type MacroBuilder = {
  /**
   * Add custom type definition, it will be written to .d.ts before macro signature.
   */
  withCustomType: (typeDefinition: string) => Omit<MacroBuilder, 'withHandler'>
  /**
   * Add signature of macro. one macro requires at least one signature.
   * @param signature function signature, like '(s: string): void'
   * @param comment comment for the signature, will be written to .d.ts
   * before the corresponding function signature.
   */
  withSignature: (signature: string, comment?: string) => MacroBuilder
  /**
   * Set the transform handler and get the macro.
   */
  withHandler: (handler: MacroHandler) => Macro
}

/**
 * Define a macro.
 * @param name the name of the macro, should be a valid identifier name.
 * @return macro builder.
 */
export function defineMacro(name: string): Omit<MacroBuilder, 'withHandler'> {
  if (!validateFnName(name))
    throw new Error(`'${name}' is not a valid macro name!`)

  const meta: MacroMeta = {
    signatures: [],
    typeDefinitions: [],
  }

  const builder: MacroBuilder = {
    withCustomType(typeDefinition) {
      meta.typeDefinitions.push(typeDefinition)
      return builder
    },
    withSignature(signature, comment) {
      if (comment) {
        comment = `/** ${comment} **/`
      }
      meta.signatures.push({ signature, comment })
      return builder
    },
    withHandler(handler) {
      if (meta.signatures.length === 0)
        throw new Error(
          `Please call .withSignature() before .withHandler() to specify at least one signature for macro '${name}'`
        )

      return {
        name,
        meta,
        apply: handler,
      }
    },
  }

  return builder
}
