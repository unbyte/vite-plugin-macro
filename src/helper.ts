import { dirname, resolve, sep } from 'path'
import { existsSync } from 'fs'
import { Helper, ImportOption } from './macro'
import { NodePath } from '@babel/traverse'
import template from '@babel/template'
import {
  File,
  ImportDeclaration,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isImportSpecifier,
  Program,
} from '@babel/types'
import { findProgramPath } from './common'

const projectDirCache = {
  root: '',
  leaf: '',
}

const projectDir: Helper['projectDir'] = (which) => {
  if (projectDirCache[which]) return projectDirCache[which]

  const paths = (require.main?.path || process.cwd()).split(sep)

  if (which === 'root') {
    for (let i = 0; i < paths.length; i++) {
      const path = paths.slice(0, i + 1).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['root'] = path)
    }
  } else {
    for (let i = paths.length; i > 0; i--) {
      const path = paths.slice(0, i).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['leaf'] = path)
    }
  }

  throw new Error('can not find project root')
}

const generateImportStmt = (imp: ImportOption) =>
  'defaultName' in imp
    ? `import ${imp.defaultName} from '${imp.moduleName}'`
    : 'localName' in imp
    ? `import { ${imp.exportName} as ${imp.localName} } from '${imp.moduleName}'`
    : 'exportName' in imp
    ? `import { ${imp.exportName} } from '${imp.moduleName}'`
    : 'namespaceName' in imp
    ? `import * as ${imp.namespaceName} from '${imp.moduleName}'`
    : `import '${imp.moduleName}'`

export function getHelper(
  filepath: string,
  ast: File
): Omit<Helper, 'forceRecollectMacros'> {
  const thisProgram = findProgramPath(ast)

  const normalizePathPattern: Helper['normalizePathPattern'] = (
    pattern,
    root = projectDir('leaf'),
    importer = filepath
  ) => {
    if (!pattern.startsWith('.') && !pattern.startsWith('/')) {
      throw new Error(
        `pattern must start with "." or "/" (relative to project root)`
      )
    }
    let base: string
    let parentDepth = 0
    const isAbsolute = pattern.startsWith('/')
    if (isAbsolute) {
      base = resolve(root)
      pattern = pattern.slice(1)
    } else {
      base = dirname(importer)
      while (pattern.startsWith('../')) {
        pattern = pattern.slice(3)
        base = resolve(base, '../')
        parentDepth++
      }
      if (pattern.startsWith('./')) {
        pattern = pattern.slice(2)
      }
    }
    return {
      normalized: pattern,
      base,
      resolveImportPath: isAbsolute
        ? (f) => `/${f}`
        : parentDepth
        ? (f) => `${'../'.repeat(parentDepth)}${f}`
        : (f) => `./${f}`,
    }
  }

  const findImported: Helper['findImported'] = (imp, program = thisProgram) => {
    // An import declaration can only be used in top-level.
    for (const path of program.get('body') as NodePath[]) {
      if (!path.isImportDeclaration()) continue
      if (!(path.node.source.value === imp.moduleName)) continue
      if ('defaultName' in imp) {
        // import defaultName from 'moduleName'
        for (const s of path.node.specifiers) {
          if (isImportDefaultSpecifier(s) && s.local.name === imp.defaultName)
            return path
        }
      } else if ('localName' in imp) {
        // import { exportName as localName } from 'moduleName'
        for (const s of path.node.specifiers) {
          if (
            isImportSpecifier(s) &&
            s.local.name === imp.localName &&
            isIdentifier(s.imported) &&
            s.imported.name === imp.exportName
          )
            return path
        }
      } else if ('exportName' in imp) {
        // import { exportName } from 'moduleName'
        for (const s of path.node.specifiers) {
          if (
            isImportSpecifier(s) &&
            isIdentifier(s.imported) &&
            s.imported.name === imp.exportName
          )
            return path
        }
      } else if ('namespaceName' in imp) {
        // import * as defaultName from 'moduleName'
        for (const s of path.node.specifiers) {
          if (
            isImportNamespaceSpecifier(s) &&
            s.local.name === imp.namespaceName
          )
            return path
        }
      } else {
        // import 'moduleName'
        return path
      }
    }
  }

  const hasImported: Helper['hasImported'] = (imp, program = thisProgram) => {
    return findImported(imp, program) !== undefined
  }

  // returns
  // - array: generated import statements
  // - path: all import statements are duplicated, returns the node path of the last one
  function normalizeImports(
    imports: ImportOption | ImportOption[],
    program: NodePath<Program>
  ) {
    if (!Array.isArray(imports)) imports = [imports]
    // remove duplicated
    const toBeImported = imports.filter((imp) => !hasImported(imp, program))
    // all import statements are duplicated, returns the node path of the last one
    if (!toBeImported.length)
      return findImported(imports[imports.length - 1], program)!
    return template.statements.ast(
      Array.from(
        new Set(toBeImported.map((imp) => generateImportStmt(imp)))
      ).join('; ')
    ) as ImportDeclaration[]
  }

  const prependImports: Helper['prependImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!Array.isArray(toBeImported)) return toBeImported
    const firstImport = (program.get('body') as NodePath[]).filter((p) =>
      p.isImportDeclaration()
    )[0]
    return (
      firstImport
        ? firstImport.insertBefore(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const appendImports: Helper['appendImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!Array.isArray(toBeImported)) return toBeImported
    const lastImport = (program.get('body') as NodePath[])
      .filter((p) => p.isImportDeclaration())
      .pop()
    return (
      lastImport
        ? lastImport.insertAfter(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const prependToBody: Helper['prependToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.unshiftContainer('body', nodes).pop() as NodePath
  }

  const appendToBody: Helper['appendToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.pushContainer('body', nodes).pop() as NodePath
  }

  const getProgram: Helper['getProgram'] = (node) => {
    if (!node) return thisProgram
    return node.findParent((p) => p.isProgram()) as NodePath<Program>
  }

  return {
    projectDir,
    normalizePathPattern,
    findImported,
    hasImported,
    prependImports,
    appendImports,
    prependToBody,
    appendToBody,
    getProgram,
  }
}
