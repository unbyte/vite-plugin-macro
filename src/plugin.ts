import type { Plugin, ViteDevServer } from 'vite'
import { ModuleNode } from 'vite'
import { mkdir, readFile, writeFile } from 'fs/promises'
import { BabelTools, Macro } from './macro'
import { parse, ParserPlugin, parseExpression } from '@babel/parser'
import * as types from '@babel/types'
import {
  CallExpression,
  File,
  ImportDeclaration,
  isCallExpression,
  isIdentifier,
  isImportDefaultSpecifier,
  isImportNamespaceSpecifier,
  isMemberExpression,
  isStringLiteral,
  Node,
} from '@babel/types'
import template from '@babel/template'
import traverse, { NodePath } from '@babel/traverse'
import { dirname, resolve } from 'path'
import { render } from 'mustache'
import generate from '@babel/generator'
import { getHelper } from './helper'
import { findInSet } from './common'

export type MacroPluginOptions = {
  /**
   * Exports macros so macros can be imported and called,
   * the value of the key will be the name of namespace(or called module).
   *
   * e.g.
   * ```typescript
   * const macroA = defineMacro('macroA')
   *   .withSignature('(...args: any[]): void')
   *   .withHandler(()=>...)
   *
   * { exports: {'@macros': {macros: [macroA]}} }
   * ```
   * Then in some .js(x) or .ts(x) file you can
   * ```typescript
   * import { macroA } from '@macros'
   * macroA(someArgs)
   * ```
   */
  exports: {
    [namespace: string]: {
      macros: Macro[]
      /**
       * Type definitions, will be written to d.ts.
       *
       * e.g.
       * ```typescript
       * { exports: { '@macros': { customTypes: `type A = string` } } }
       * ```
       * will generate
       * ```typescript
       * declare module '@macros' {
       *   type A = string
       * }
       * ```
       */
      customTypes?: string
    }
  }
  /**
   * The path of the automatically generated type declaration file.
   */
  dtsPath: string
  /**
   * The max recursion for applying macros, default to 5.
   *
   * After reached the maxRecursion, plugin will throw out an error.
   *
   * It's usually caused by forgetting to remove the call expression
   * in macros.
   */
  maxRecursion?: number
  /**
   * Babel plugins to be applied during parsing.
   *
   * By default 'typescript' and 'jsx' are included and cannot be removed.
   */
  parserPlugins?: ParserPlugin[]
}

export type MacroPluginServerHelper = {
  /**
   * Invalidate caches of modules that import macros
   * from all namespaces or specific namespace,
   * only available in dev mode.
   *
   * This is usually used when, the macro needs to be re-expanded due to
   * the change of external conditions, while the modules that call
   * the macro cannot be automatically reloaded because their dependencies
   * in the module dependency graph have no change so Vite always caches them.
   */
  invalidateCache: (namespace?: string) => void
  /**
   * Find modules that import macros from all namespaces or specific namespace,
   * only available in dev mode.
   */
  findImporters: (namespace?: string) => Set<ModuleNode>
}

export type MacroPluginServerHook = (
  server: ViteDevServer,
  helpers: MacroPluginServerHelper
) => (() => void) | void | Promise<(() => void) | void>

export type MacroPlugin = MacroPluginOptions & {
  /**
   * The name of plugin.
   */
  name: string
  /**
   * Vite plugin options.
   */
  hooks?: Omit<Plugin, 'name' | 'enforce' | 'apply' | 'configureServer'> & {
    configureServer?: MacroPluginServerHook
  }
}

type NamespacedMacros = { [namespace: string]: Macro[] }

type NamespacedCustomTypes = { [namespace: string]: string | undefined }

type NormalizedMacroPluginOptions = {
  dtsPath: string
  namespaces: string[]
  macros: NamespacedMacros
  maxRecursion: number
  parserPlugins: ParserPlugin[]
  customTypes: NamespacedCustomTypes
}

function normalizeOption(
  raw: MacroPluginOptions
): NormalizedMacroPluginOptions {
  const { dtsPath, exports, maxRecursion, parserPlugins } = raw
  const namespaces = Object.keys(exports)
  const macros = Object.create(null)
  const customTypes = Object.create(null)
  namespaces.forEach((ns) => {
    macros[ns] = exports[ns].macros
    customTypes[ns] = exports[ns].customTypes
  })
  return {
    dtsPath,
    namespaces,
    macros,
    customTypes,
    maxRecursion: maxRecursion && maxRecursion > 0 ? maxRecursion : 5,
    parserPlugins: ['typescript', 'jsx', ...(parserPlugins || [])],
  }
}

type MacroPluginContext = {
  dev: boolean
}

/**
 * Define the macro plugin. It can be used as vite plugin directly.
 * @param plugin plugin options.
 */
export function defineMacroPlugin(plugin: MacroPlugin): Plugin {
  const {
    name,
    hooks: {
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    } = {},
    ...options
  } = plugin

  const {
    dtsPath,
    namespaces,
    macros,
    maxRecursion,
    parserPlugins,
    customTypes,
  } = normalizeOption(options)

  throwErrorIfConflict(macros)

  const ctx: MacroPluginContext = {
    dev: false,
  }

  return {
    name,
    enforce: 'pre',
    configResolved(config) {
      if (config.env.DEV) ctx.dev = true
      generateDts(dtsPath, {
        namespaces,
        macros,
        customTypes: customTypes,
      }).then()
      // hook
      return configResolved?.(config)
    },
    resolveId(id, importer, options, ssr) {
      if (namespaces.includes(id)) return id
      // hook
      return resolveId?.bind(this)(id, importer, options, ssr)
    },
    load(id, ssr) {
      if (namespaces.includes(id)) return 'export {}'
      // hook
      return load?.bind(this)(id, ssr)
    },
    transform(code, id, ssr) {
      if (/\.[jt]sx?$/.test(id)) {
        const ast = parse(code, {
          sourceType: 'module',
          plugins: parserPlugins,
        })

        const importedMacros = getImportedMacros(
          ast,
          namespaces,
          // Note: keep import statements in dev mode so can invalidate macro modules
          ctx.dev /* keep import stmt */
        )

        if (!importedMacros.collect().size) return

        const applyCtx: ApplyContext = {
          code,
          filepath: id,
          ast,
          macroFinder: getMacroFinder(importedMacros, macros),
        }

        let loopCount = 0
        while (loopCount < maxRecursion) {
          const { applied, recollectMacros } = applyMacros(applyCtx)
          if (!applied) break
          if (recollectMacros) importedMacros.collect()

          loopCount++
          if (loopCount === maxRecursion)
            throw new Error(
              `Reached the maximum recursion, please check macros applied in file ${id}`
            )
        }

        return generate(ast, {
          retainLines: true,
        }).code
      }
      // hook
      return transform?.bind(this)(code, id, ssr)
    },
    configureServer(server) {
      // hook
      function findImporters(ns?: string) {
        const result = new Set<ModuleNode>()
        const moduleIDs = ns ? [ns] : namespaces
        moduleIDs.forEach((id) => {
          server.moduleGraph
            .getModuleById(`/@id/${id}`)
            ?.importers.forEach((m) => result.add(m))
        })
        return result
      }

      function invalidateCache(ns?: string) {
        findImporters(ns)?.forEach((m) =>
          server.moduleGraph.invalidateModule(m)
        )
      }

      configureServer?.(server, { findImporters, invalidateCache })
    },
    ...otherHooks,
  }
}

function throwErrorIfConflict(macros: NormalizedMacroPluginOptions['macros']) {
  Object.keys(macros).forEach((ns) => {
    const mem = Object.create(null)
    macros[ns].forEach((m) => {
      if (mem[m.name]) {
        throw new Error(
          `Error when loading macros: a macro with name '${m.name}' in '${ns}' already existed`
        )
      }
      mem[m.name] = 1
    })
  })
}

type ImportedMacro =
  | {
      importedAsNamespace: true
      namespace: string
      local: string
    }
  | {
      importedAsNamespace: false
      namespace: string
      local: string
      origin: string
    }

type ImportedMacrosCollector = Set<ImportedMacro> & {
  collect: () => ImportedMacrosCollector
}

function getImportedMacros(
  ast: Node,
  namespaces: string[],
  keepImportStmt = false
): ImportedMacrosCollector {
  const importedMacros: ImportedMacrosCollector = new Set() as any

  importedMacros.collect = () => {
    {
      const importPaths: NodePath<ImportDeclaration>[] = []
      traverse(ast, {
        Declaration(path) {
          if (!path.isImportDeclaration()) return
          if (!namespaces.includes(path.node.source.value)) return

          // case - import 'a'
          if (!path.node.specifiers.length) return

          path.node.specifiers.forEach((s) => {
            importedMacros.add(
              // case - import a from 'a'
              // or
              // case - import * as a from 'a'
              isImportDefaultSpecifier(s) || isImportNamespaceSpecifier(s)
                ? {
                    importedAsNamespace: true,
                    namespace: path.node.source.value,
                    local: s.local.name,
                  }
                : // case - import { a as a } from 'a'
                  {
                    importedAsNamespace: false,
                    namespace: path.node.source.value,
                    local: s.local.name,
                    // imported !== existed
                    origin: isIdentifier(s.imported)
                      ? s.imported.name
                      : s.imported.value,
                  }
            )
          })
          importPaths.push(path)
        },
      })
      if (keepImportStmt) {
        importPaths.forEach((p) =>
          p.replaceWith(
            template.statement.ast(`import '${p.node.source.value}'`)
          )
        )
      } else {
        importPaths.forEach((p) => p.remove())
      }
    }
    return importedMacros
  }

  return importedMacros
}

async function generateDts(
  targetPath: string,
  {
    namespaces,
    macros,
    customTypes,
  }: Pick<NormalizedMacroPluginOptions, 'namespaces' | 'macros' | 'customTypes'>
) {
  await mkdir(dirname(targetPath), { recursive: true })
  await writeFile(
    targetPath,
    render(
      (await readFile(resolve(__dirname, './client.d.ts.tpl'))).toString(),
      {
        namespaces: namespaces.map((ns) => ({
          module: ns,
          macros: macros[ns].map((m) => ({
            macroScopeTypes: m.meta.typeDefinitions,
            signature: m.meta.signatures,
            name: m.name,
          })),
          moduleScopeTypes: customTypes[ns],
        })),
      }
    )
  )
}

// find macros to be applied from call expr, returns a
// - Macro: has origin
// - string: has origin name but no origin macro found
// - undefined: no origin name found
type MacroFinder = (
  callee: CallExpression['callee']
) => Macro | string | undefined

function getMacroFinder(
  importedMacros: Set<ImportedMacro>,
  macros: NamespacedMacros
): MacroFinder {
  return (callee) => {
    if (isMemberExpression(callee)) {
      const ns = callee.object
      if (isIdentifier(ns)) {
        const maybeMacro = findInSet(
          importedMacros,
          (m) => m.importedAsNamespace && m.local === ns.name
        )
        if (!maybeMacro) return
        const method = callee.property
        if (isIdentifier(method)) {
          // case - namespace.method()
          return (
            macros[maybeMacro.namespace].find((m) => m.name === method.name) ||
            method.name
          )
        } else if (isStringLiteral(method)) {
          // case - namespace['method']()
          return (
            macros[maybeMacro.namespace].find((m) => m.name === method.value) ||
            method.value
          )
        }
      }
    } else if (isIdentifier(callee)) {
      // case - method()
      const maybeMacro = findInSet(
        importedMacros,
        (m) => !m.importedAsNamespace && m.local === callee.name
      )
      if (!maybeMacro) return
      return (
        macros[maybeMacro.namespace].find(
          (m) => m.name === (maybeMacro as { origin: string }).origin
        ) || callee.name
      )
    }
  }
}

const BABEL_TOOLS: BabelTools = Object.freeze({
  template,
  traverse,
  types,
  parse,
  parseExpression,
})

type ApplyContext = {
  code: string
  filepath: string
  ast: File

  macroFinder: MacroFinder
}

type ApplyResult = {
  applied: boolean
  recollectMacros: boolean
}

function applyMacros({
  code,
  filepath,
  ast,
  macroFinder,
}: ApplyContext): ApplyResult {
  const result: ApplyResult = {
    applied: false,
    recollectMacros: false,
  }

  const helper = Object.freeze({
    forceRecollectMacros: () => (result.recollectMacros = true),
    ...getHelper(filepath, ast),
  })

  traverse(ast, {
    Expression(path) {
      if (isCallExpression(path.node)) {
        const macroToApply = macroFinder(path.node.callee)
        if (!macroToApply) return

        if (typeof macroToApply === 'string')
          throw new Error(`macro '${macroToApply}' not found in ${filepath}`)

        try {
          macroToApply.apply(
            {
              code,
              filepath,
              path,
              args: path.node.arguments,
            },
            BABEL_TOOLS,
            helper
          )
        } catch (e) {
          throw new Error(
            `Error when apply macro ${macroToApply.name} in ${filepath}: ${e}`
          )
        } finally {
          result.applied = true
        }
      }
    },
  })

  return result
}
