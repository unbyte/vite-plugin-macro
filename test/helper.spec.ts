import { getHelper } from '../src/helper'
import { File } from '@babel/types'
import { getAST, matchCodeSnapshot } from './testutils'
import template from '@babel/template'
import { ImportOption } from '../src'

describe('helper', () => {
  let ast: File
  let helper: ReturnType<typeof getHelper>

  beforeEach(() => {
    ast = getAST(`
  import { a as _a } from 'a'
  import na from 'a'
  import * as _na from 'a'
  import { a } from 'a'
  import 'a'
  const b = a(1)
  `)
    helper = getHelper('/workspace/project-a/src/test.ts', ast)
  })

  it('projectDir() should work', () => {
    const leaf = helper.projectDir('leaf')
    const root = helper.projectDir('root')
    expect(leaf).not.toBeUndefined()
    expect(root).not.toBeUndefined()
    // in this project it should be equal
    expect(leaf).toBe(root)
  })

  it('getProgram() should work', () => {
    expect(helper.getProgram()).not.toBeUndefined()
  })

  it('findImported() should work', () => {
    const imps: ImportOption[] = [
      { moduleName: 'a', exportName: 'a' },
      { moduleName: 'a', exportName: 'a', localName: '_a' },
      { moduleName: 'a' },
      { moduleName: 'a', defaultName: 'na' },
      { moduleName: 'a', namespaceName: '_na' },
    ]
    imps.forEach((item) =>
      expect(helper.findImported(item)).not.toBeUndefined()
    )
    expect(
      helper.findImported({ moduleName: 'a', exportName: 'b' })
    ).toBeUndefined()
  })

  it('hasImported() should work', () => {
    expect(helper.hasImported({ moduleName: 'a', exportName: 'a' })).toBe(true)
    expect(helper.hasImported({ moduleName: 'a', exportName: 'b' })).toBe(false)
  })

  it('prependImports() should work', () => {
    expect(helper.prependImports({ moduleName: 'a', exportName: 'a' })).toBe(
      helper.findImported({ moduleName: 'a', exportName: 'a' })
    )
    expect(
      helper.prependImports({
        moduleName: 'a',
        exportName: 'b',
      })
    ).not.toBeUndefined()
    matchCodeSnapshot(ast)
  })

  it('appendImports() should work', () => {
    expect(helper.appendImports({ moduleName: 'a', exportName: 'a' })).toBe(
      helper.findImported({ moduleName: 'a', exportName: 'a' })
    )
    expect(
      helper.appendImports({
        moduleName: 'a',
        exportName: 'b',
      })
    ).not.toBeUndefined()
    matchCodeSnapshot(ast)
  })

  it('prependToBody() should work', () => {
    helper.prependToBody(template.statement.ast(`console.log('hello world')`))
    matchCodeSnapshot(ast)
  })

  it('appendToBody() should work', () => {
    helper.appendToBody(template.statement.ast(`console.log('hello world')`))
    matchCodeSnapshot(ast)
  })

  it('normalizePathPattern() should work', () => {
    const { normalized, resolveImportPath } = helper.normalizePathPattern(
      '../assets/*.css',
      '/workspace/a',
      '/workspace/a/src/test.ts'
    )
    expect(normalized).toBe('assets/*.css')
    expect(resolveImportPath('assets/hello.css')).toBe('../assets/hello.css')
  })
})
