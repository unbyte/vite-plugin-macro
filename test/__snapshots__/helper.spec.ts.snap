// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper appendImports() should work 1`] = `
"import { a as _a } from 'a';
import na from 'a';
import * as _na from 'a';
import { a } from 'a';
import 'a';
import { b } from 'a';
const b = a(1);"
`;

exports[`helper appendToBody() should work 1`] = `
"import { a as _a } from 'a';
import na from 'a';
import * as _na from 'a';
import { a } from 'a';
import 'a';
const b = a(1);
console.log('hello world');"
`;

exports[`helper prependImports() should work 1`] = `
"import { b } from 'a';
import { a as _a } from 'a';
import na from 'a';
import * as _na from 'a';
import { a } from 'a';
import 'a';
const b = a(1);"
`;

exports[`helper prependToBody() should work 1`] = `
"console.log('hello world');
import { a as _a } from 'a';
import na from 'a';
import * as _na from 'a';
import { a } from 'a';
import 'a';
const b = a(1);"
`;
